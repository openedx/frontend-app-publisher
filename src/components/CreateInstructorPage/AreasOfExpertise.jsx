import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';

import RenderInputTextField from '../RenderInputTextField';
import RemoveButton from '../../components/RemoveButton';


class AreasOfExpertise extends React.Component {
  constructor(props) {
    super(props);

    this.handleRemove = this.handleRemove.bind(this);
  }

  handleRemove(index) {
    this.props.fields.remove(index);
  }

  render() {
    const { fields } = this.props;

    return (
      <div className="areas-of-expertise my-4">
        <button
          type="button"
          className="btn btn-light js-add-button"
          onClick={() => fields.push({})}
        >
          Add area of expertise
        </button>
        <ul className="list-group p-0 m-0">
          {fields.map((expertise, index) => (
            <li className="area-of-expertise list-group-item row d-flex align-items-center px-0 mx-0" key={expertise}>
              <div className="col-10">
                <Field
                  name={`${expertise}.value`}
                  component={RenderInputTextField}
                  type="text"
                  label={
                    <React.Fragment>
                      Area of expertise
                      <span className="required" aria-hidden>*</span>
                    </React.Fragment>
                  }
                  required
                />
              </div>
              <RemoveButton
                label="Remove area of expertise"
                onRemove={this.handleRemove}
                targetFieldNumber={index}
              />
            </li>
          ))}
        </ul>
      </div>
    );
  }
}

AreasOfExpertise.propTypes = {
  fields: PropTypes.shape({
    remove: PropTypes.func,
  }).isRequired,
};

export default AreasOfExpertise;
