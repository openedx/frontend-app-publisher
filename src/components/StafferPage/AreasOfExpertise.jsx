import React from 'react';
import PropTypes from 'prop-types';
import { Field, FieldArray } from 'redux-form';

import RenderInputTextField from '../RenderInputTextField';
import RemoveButton from '../RemoveButton';
import FieldLabel from '../FieldLabel';

class AreasOfExpertise extends React.Component {
  constructor(props) {
    super(props);

    this.handleRemove = this.handleRemove.bind(this);
  }

  handleRemove(index) {
    this.props.fields.remove(index);
  }

  render() {
    const { fields } = this.props;

    return (
      <div className="areas-of-expertise mb-3">
        <ul className="list-group p-0 m-0 container-fluid">
          {fields.map((expertise, index) => (
            <li className="area-of-expertise list-group-item row d-flex align-items-center px-0 mx-0" key={expertise}>
              <div className="col-11">
                <Field
                  name={`${expertise}.value`}
                  component={RenderInputTextField}
                  type="text"
                  label={<FieldLabel text="Area of expertise" required />}
                  required
                />
              </div>
              <input
                name={`${expertise}.id`}
                type="hidden"
              />
              <RemoveButton
                className="col-1"
                label="Remove area of expertise"
                onRemove={this.handleRemove}
                targetFieldNumber={index}
              />
            </li>
          ))}
        </ul>
        <button
          type="button"
          className="btn btn-outline-primary js-add-button mt-2"
          onClick={() => fields.push({})}
        >
          Add area of expertise
        </button>
      </div>
    );
  }
}

AreasOfExpertise.propTypes = {
  fields: PropTypes.instanceOf(FieldArray).isRequired,
};

export default AreasOfExpertise;
