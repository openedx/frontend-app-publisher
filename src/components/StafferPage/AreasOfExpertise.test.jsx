import React from 'react';
import { shallow } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';

import AreasOfExpertise from './AreasOfExpertise';

/*
*  Disable console errors for this test file so that we don't receive warnings
*  about fields being an array rather than an object. This prop change is
*  intentional as redux-form field arrays treat fields strangely:
*  'The fields object is a "pseudo-array", in that it has many of the same
*  properties and methods as a javascript Array, providing both reading and
*  writing functionality.'
*/
jest.spyOn(global.console, 'error').mockImplementation(() => jest.fn());

describe('Areas Of Expertise', () => {
  it('renders correctly with no fields', () => {
    const component = shallow(<AreasOfExpertise fields={[]} />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('renders correctly when given fields', () => {
    const component = shallow(<AreasOfExpertise fields={[{}]} />);
    expect(shallowToJson(component)).toMatchSnapshot();
  });

  it('adds fields when the add button is pushed', () => {
    const fields = [{}];
    const component = shallow(<AreasOfExpertise fields={fields} />);
    expect(fields.length).toEqual(1);

    component.find('.js-add-button').simulate('click');
    expect(fields.length).toEqual(2);
  });
});
