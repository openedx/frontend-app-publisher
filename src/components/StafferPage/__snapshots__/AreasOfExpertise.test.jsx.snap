// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Areas Of Expertise renders correctly when given fields 1`] = `
<div
  className="areas-of-expertise mb-3"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  >
    <li
      className="area-of-expertise list-group-item row d-flex align-items-center px-0 mx-0"
      key="[object Object]"
    >
      <div
        className="col-11"
      >
        <Field
          component={[Function]}
          label={
            <FieldLabel
              className=""
              extraText=""
              helpText=""
              id={null}
              optional={false}
              required={true}
              text="Area of expertise"
            />
          }
          name="[object Object].value"
          required={true}
          type="text"
        />
      </div>
      <input
        name="[object Object].id"
        type="hidden"
      />
      <RemoveButton
        className="col-1"
        label="Remove area of expertise"
        onRemove={[Function]}
        targetFieldNumber={0}
      />
    </li>
  </ul>
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    onClick={[Function]}
    type="button"
  >
    Add area of expertise
  </button>
</div>
`;

exports[`Areas Of Expertise renders correctly with no fields 1`] = `
<div
  className="areas-of-expertise mb-3"
>
  <ul
    className="list-group p-0 m-0 container-fluid"
  />
  <button
    className="btn btn-outline-primary js-add-button mt-2"
    onClick={[Function]}
    type="button"
  >
    Add area of expertise
  </button>
</div>
`;
